# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [1]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [2]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [3]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [4]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [5]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [6]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [7]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [8]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [9]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [10]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [11]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [12]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [13]:
df['Season'].value_counts()

2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: Season, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [14]:
#My code here

Totalseasons=df['Season'].nunique()

print(Totalseasons)

#By first defining this df with the unique function and then using the print function i know that there is a total of 32 seasons in total.

32


**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [15]:
#My code here
df['Wteam'].value_counts()

#Here i can see that it is team 1181 that has most wins by 819. I can also do it on another way (see code below) that is more explicit for the reader

1181    819
1242    804
1246    765
1314    761
1112    746
       ... 
1101     18
1446     14
1118      6
1289      6
1327      3
Name: Wteam, Length: 364, dtype: int64

In [16]:
# Calculate the number of wins for each team
team_wins = df['Wteam'].value_counts()

# Find the team with the most wins
team_with_most_wins = team_wins.idxmax()
most_wins = team_wins.max()

print('Team most wins:')
print("Team:", team_with_most_wins)
print("Wins:", most_wins)

Team most wins:
Team: 1181
Wins: 819


# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [17]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [18]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [19]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [20]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [21]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [22]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [23]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [24]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [25]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [26]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [27]:
#write your code here
#As i have to sort different things i have created some comments that makes it easier for the reader to see what is going on: 

#Df sorted by season
sort_season = df.sort_values('Season', ascending=False)

#Df sorted by Winning Team
sort_wteam = df.sort_values('Wteam',ascending=False)

#Df sorted by Winning Score
sort_wscore = df.sort_values('Wscore', ascending=False)

df10 = sort_season.iloc[100:200]
df11 = sort_wteam.iloc[100:200]
df12 = sort_wscore.iloc[100:200]

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [28]:
print('highest score =', df12.iloc[0,3])
print('Team =', df12.iloc[0,2])
print('season =', df12.iloc[0,0])


highest score = 130
Team = 1258
season = 1988


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [29]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [30]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [31]:
# Step 1: Create a new column 'ScoreDifference'
df['scoreDif'] = abs(df['Wscore'] - df['Lscore'])

# Step 2: Calculate the average 'ScoreDifference' for all games
average_diff = df['scoreDif'].mean()

# Step 3: Filter the DataFrame based on the condition 'ScoreDifference' > average_diff using loc
filtered_df = df[df['scoreDif'] > average_diff]

# Print the filtered DataFrame
filtered_df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  scoreDif
0         1985      20   1228      81   1328      64    N      0        17
3         1985      25   1165      70   1432      54    H      0        16
6         1985      25   1228      64   1226      44    N      0        20
8         1985      25   1260      98   1133      80    H      0        18
10        1985      25   1307     103   1288      71    H      0        32
...        ...     ...    ...     ...    ...     ...  ...    ...       ...
145280    2016     131   1401      71   1261      38    N      0        33
145282    2016     131   1433      76   1172      54    N      0        22
145284    2016     132   1114      70   1419      50    N      0        20
145285    2016     132   1163      72   1272      58    N      0        14
145288    2016     132   1386      87   1433      74    N      0        13

[57227 rows x 9 columns]

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [32]:
#My code here: 
filtered_df.loc[filtered_df['scoreDif'].idxmax()]

Season      1996
Daynum        48
Wteam       1409
Wscore       141
Lteam       1341
Lscore        50
Wloc           H
Numot          0
scoreDif      91
Name: 44653, dtype: object

This shows me that it is in season 1996 it was team 1409 that the highest scoredifference with 91. The losing team was 1341. 

# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [33]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [34]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: Wloc, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [35]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [36]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [37]:
#My code here
df.groupby('Season').agg({'Wscore': 'mean'})

Wscore
Season           
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958

Above is all the seasons displayed with the average winning score from each season from 1985 to 2016. 

**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [38]:
#My code here
df.groupby(['Wteam'])['Wscore', 'Season'].max()

/var/folders/_8/b5t0w1254n3crscm8b3vmy3m0000gn/T/ipykernel_2321/1645323925.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df.groupby(['Wteam'])['Wscore', 'Season'].max()


Wscore  Season
Wteam                
1101       95    2016
1102      111    2016
1103      109    2016
1104      114    2016
1105      114    2016
...       ...     ...
1460      136    2016
1461      112    2016
1462      125    2016
1463      105    2016
1464      115    2016

[364 rows x 2 columns]

In [39]:
#My code here
df.groupby('Wteam').agg({'Wscore': 'max', 'Season': 'max'})

Wscore  Season
Wteam                
1101       95    2016
1102      111    2016
1103      109    2016
1104      114    2016
1105      114    2016
...       ...     ...
1460      136    2016
1461      112    2016
1462      125    2016
1463      105    2016
1464      115    2016

[364 rows x 2 columns]

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [40]:
#my code here
df.groupby(['Season', 'Wteam']).agg({'Wscore': 'mean'}).reset_index().sort_values('Wscore', ascending=False).groupby('Season').first()

Wteam      Wscore
Season                   
1985     1328   92.800000
1986     1109   91.200000
1987     1380   95.875000
1988     1258  111.750000
1989     1258  117.315789
1990     1258  126.347826
1991     1380  112.312500
1992     1380   99.642857
1993     1380  101.875000
1994     1380  106.583333
1995     1206  102.833333
1996     1206  103.600000
1997     1254   94.142857
1998     1395  101.153846
1999     1317   95.750000
2000     1395   95.647059
2001     1395   99.315789
2002     1242   92.888889
2003     1395   88.222222
2004     1194   92.142857
2005     1311   91.333333
2006     1190   89.533333
2007     1440   99.400000
2008     1440   94.500000
2009     1377   96.200000
2010     1440   95.000000
2011     1449   90.130435
2012     1370   87.400000
2013     1377   86.500000
2014     1322   92.600000
2015     1322   90.733333
2016     1146   97.142857

**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [41]:
#My code here
New_df7 = df.groupby(['Season', 'Wteam']).size().rename('Count')
print('Number of wins for each team for each season:')
print(New_df7)

Number of wins for each team for each season:
Season  Wteam
1985    1102      5
        1103      9
        1104     21
        1106     10
        1108     19
                 ..
2016    1460     20
        1461     12
        1462     27
        1463     21
        1464      9
Name: Count, Length: 10172, dtype: int64


**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [42]:
# Write your code here

team_win_count = df.groupby(['Season', 'Wteam']).size().reset_index(name='Count')
teams_with_most_wins = team_win_count.groupby('Season')['Count'].idxmax().reset_index()
teams_with_most_wins = team_win_count.loc[teams_with_most_wins['Count']]

print(teams_with_most_wins)

      Season  Wteam  Count
217     1985   1385     27
342     1986   1181     32
818     1987   1424     33
861     1988   1112     31
1323    1989   1328     28
1551    1990   1247     29
1739    1991   1116     30
2088    1992   1181     28
2423    1993   1231     28
2665    1994   1163     27
2928    1995   1116     26
3357    1996   1269     31
3638    1997   1242     31
3943    1998   1242     33
4201    1999   1181     32
4709    2000   1409     29
4829    2001   1181     29
5121    2002   1153     30
5456    2003   1166     29
5981    2004   1390     29
6165    2005   1228     32
6452    2006   1181     30
6841    2007   1242     30
7209    2008   1272     33
7552    2009   1272     31
7870    2010   1242     32
8217    2011   1242     32
8565    2012   1246     32
8878    2013   1211     30
9462    2014   1455     33
9609    2015   1246     34
9955    2016   1242     29


**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [43]:
# Write your code here
df.groupby(['Season', 'Lteam']).agg({'Lscore': 'mean'}).reset_index().sort_values(['Season', 'Lscore'], ascending=[True, False]).groupby('Season').first()

Lteam      Lscore
Season                   
1985     1260   81.000000
1986     1109   90.157895
1987     1424   88.000000
1988     1379   89.600000
1989     1258   99.200000
1990     1258  117.800000
1991     1258   95.866667
1992     1258   86.615385
1993     1246   85.666667
1994     1407   91.785714
1995     1206   87.100000
1996     1407   88.562500
1997     1242   94.000000
1998     1254   89.400000
1999     1373   84.600000
2000     1390   79.666667
2001     1181   83.750000
2002     1332   82.000000
2003     1400   81.166667
2004     1112   81.666667
2005     1310   82.888889
2006     1144   80.777778
2007     1440   93.526316
2008     1314   79.000000
2009     1440   84.625000
2010     1440   83.263158
2011     1440   79.307692
2012     1314   75.000000
2013     1233   77.230769
2014     1140   80.090909
2015     1233   75.625000
2016     1453   81.416667

# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [44]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season      1985
Daynum        20
Wteam       1228
Wscore        81
Lteam       1328
Lscore        64
Wloc           N
Numot          0
scoreDif      17
Name: 0, dtype: object
Season      1985
Daynum        25
Wteam       1106
Wscore        77
Lteam       1354
Lscore        70
Wloc           H
Numot          0
scoreDif       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [45]:
# Write your code here
df['HighScoringGame'] = ['Yes' if score > 100 else 'No' for score in df['Wscore']]
print(df)

        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  scoreDif  \
0         1985      20   1228      81   1328      64    N      0        17   
1         1985      25   1106      77   1354      70    H      0         7   
2         1985      25   1112      63   1223      56    H      0         7   
3         1985      25   1165      70   1432      54    H      0        16   
4         1985      25   1192      86   1447      74    H      0        12   
...        ...     ...    ...     ...    ...     ...  ...    ...       ...   
145284    2016     132   1114      70   1419      50    N      0        20   
145285    2016     132   1163      72   1272      58    N      0        14   
145286    2016     132   1246      82   1401      77    N      1         5   
145287    2016     132   1277      66   1345      62    N      0         4   
145288    2016     132   1386      87   1433      74    N      0        13   

       HighScoringGame  
0                   No  
1            

In [46]:
#Then i could do like this to see the winning score over 100. 
high_scoring_games = df[df['Wscore'] > 100]
print(high_scoring_games)

        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  scoreDif  \
10        1985      25   1307     103   1288      71    H      0        32   
25        1985      26   1196     106   1416      55    H      0        51   
35        1985      26   1286     109   1186      64    H      0        45   
42        1985      26   1375     104   1126      68    H      0        36   
50        1985      26   1439     102   1330      76    H      0        26   
...        ...     ...    ...     ...    ...     ...  ...    ...       ...   
144999    2016     124   1267     108   1379     106    H      1         2   
145135    2016     128   1266     101   1385      93    N      0         8   
145148    2016     128   1424     108   1102     102    N      3         6   
145218    2016     130   1163     104   1153      97    N      4         7   
145245    2016     130   1372     104   1223      68    N      0        36   

       HighScoringGame  
10                 Yes  
25           

**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [47]:
#My code here
from collections import defaultdict

total_games = defaultdict(int)

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    # Extract the winning and losing teams
    winning_team = row['Wteam']
    losing_team = row['Lteam']

    # Update the count for the winning team
    total_games[winning_team] += 1

    # Update the count for the losing team
    total_games[losing_team] += 1

# Sort the teams alphabetically
sorted_teams = sorted(total_games.items())

print("Total number of games played by each team:")
for team, games in sorted_teams:
    print(team, games)

Total number of games played by each team:
1101 76
1102 840
1103 910
1104 975
1105 447
1106 855
1107 512
1108 866
1109 169
1110 910
1111 876
1112 981
1113 937
1114 910
1115 520
1116 980
1117 891
1118 27
1119 868
1120 936
1121 175
1122 903
1123 906
1124 906
1125 506
1126 906
1127 442
1128 106
1129 907
1130 951
1131 945
1132 895
1133 949
1134 200
1135 847
1136 240
1137 912
1138 728
1139 923
1140 965
1141 872
1142 597
1143 947
1144 850
1145 934
1146 267
1147 698
1148 844
1149 824
1150 942
1151 899
1152 865
1153 970
1154 820
1155 950
1156 900
1157 784
1158 720
1159 887
1160 928
1161 905
1162 831
1163 977
1164 893
1165 833
1166 957
1167 260
1168 891
1169 724
1170 660
1171 829
1172 879
1173 960
1174 913
1175 890
1176 526
1177 942
1178 920
1179 914
1180 930
1181 1013
1182 924
1183 871
1184 863
1185 900
1186 849
1187 861
1188 207
1189 481
1190 891
1191 928
1192 908
1193 942
1194 647
1195 273
1196 981
1197 896
1198 812
1199 949
1200 922
1201 947
1202 856
1203 933
1204 862
1205 405
1206 938
1207

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [48]:
#my code here
def print_max_score_diff(row):
    season = row['Season']
    index = row.name
    winning_team = row['Wteam']
    losing_team = row['Lteam']
    score_diff = row['scoreDif']
    print(f"Season: {season}, Game Number: {index}, Winning Team: {winning_team}, Losing Team: {losing_team}, Score Difference: {score_diff}")

max_score_diff_games = df.groupby('Season')['scoreDif'].idxmax()
df.loc[max_score_diff_games].apply(print_max_score_diff, axis=1)

Season: 1985, Game Number: 236, Winning Team: 1361, Losing Team: 1288, Score Difference: 60
Season: 1986, Game Number: 4731, Winning Team: 1314, Losing Team: 1264, Score Difference: 84
Season: 1987, Game Number: 8240, Winning Team: 1155, Losing Team: 1118, Score Difference: 73
Season: 1988, Game Number: 12046, Winning Team: 1328, Losing Team: 1147, Score Difference: 68
Season: 1989, Game Number: 16677, Winning Team: 1242, Losing Team: 1135, Score Difference: 70
Season: 1990, Game Number: 19502, Winning Team: 1181, Losing Team: 1217, Score Difference: 76
Season: 1991, Game Number: 25161, Winning Team: 1163, Losing Team: 1148, Score Difference: 68
Season: 1992, Game Number: 27997, Winning Team: 1116, Losing Team: 1126, Score Difference: 82
Season: 1993, Game Number: 33858, Winning Team: 1328, Losing Team: 1197, Score Difference: 81
Season: 1994, Game Number: 36404, Winning Team: 1228, Losing Team: 1152, Score Difference: 69
Season: 1995, Game Number: 39858, Winning Team: 1246, Losing Tea

236       None
4731      None
8240      None
12046     None
16677     None
19502     None
25161     None
27997     None
33858     None
36404     None
39858     None
44653     None
49033     None
52600     None
57618     None
60616     None
65999     None
70737     None
74791     None
79181     None
83372     None
89542     None
94507     None
97737     None
103723    None
108405    None
114015    None
118699    None
124057    None
131373    None
134590    None
140040    None
dtype: object

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [49]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [52]:
#My code here
wins = df.groupby('Wteam').size().rename('Wins')
losses = df.groupby('Lteam').size().rename('Losses')
total_games = wins.add(losses, fill_value=0)

print("Total number of games played by each team:")
print(total_games)

Total number of games played by each team:
Wteam
1101     76
1102    840
1103    910
1104    975
1105    447
       ... 
1460    827
1461    914
1462    954
1463    838
1464    856
Length: 364, dtype: int64


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [53]:
# Write your code here
df['scoreDif'] = df['Wscore'] - df['Lscore']
games_with_highest_score_diff = df.loc[df.groupby('Season')['scoreDif'].idxmax()]

print(games_with_highest_score_diff[['Season', 'Wteam', 'Lteam', 'scoreDif']])

        Season  Wteam  Lteam  scoreDif
236       1985   1361   1288        60
4731      1986   1314   1264        84
8240      1987   1155   1118        73
12046     1988   1328   1147        68
16677     1989   1242   1135        70
19502     1990   1181   1217        76
25161     1991   1163   1148        68
27997     1992   1116   1126        82
33858     1993   1328   1197        81
36404     1994   1228   1152        69
39858     1995   1246   1404        74
44653     1996   1409   1341        91
49033     1997   1278   1106        80
52600     1998   1395   1410        66
57618     1999   1268   1317        75
60616     2000   1261   1212        75
65999     2001   1196   1197        75
70737     2002   1328   1183        59
74791     2003   1403   1311        72
79181     2004   1328   1115        70
83372     2005   1403   1421        64
89542     2006   1400   1341        72
94507     2007   1401   1212        74
97737     2008   1182   1224        70
103723    2009   1314   1

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [54]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [ ]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [ ]:
type(df['Wscore'])

pandas.core.series.Series

In [ ]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [ ]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  scoreDif  \
0    1985      20   1228      81   1328      64    N      0        17   
1    1985      25   1106      77   1354      70    H      0         7   
2    1985      25   1112      63   1223      56    H      0         7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [ ]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  scoreDif  \
0    1985      20   1228      81   1328      64    N      0        17   
1    1985      25   1106      77   1354      70    H      0         7   
2    1985      25   1112      63   1223      56    H      0         7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [ ]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
scoreDif           0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y